daemon/graphdriver/devmapper/deviceset.go

```diff
@@ -1503,7 +1718,92 @@ func (devices *DeviceSet) DeleteDevice(hash string) error {
 	devices.Lock()
 	defer devices.Unlock()
 
-	return devices.deleteDevice(info)
+	// If mountcount is not zero, that means devices is still in use
+	// or has not been Put() properly. Fail device deletion.
+
+	if info.mountCount != 0 {
+		return fmt.Errorf("devmapper: Can't delete device %v as it is still mounted. mntCount=%v", info.Hash, info.mountCount)
+	}
+
+	return devices.deleteDevice(info, syncDelete)
+}
+
+// Either sleep for 30 seconds or wait for command from parent.
+// It returns true if one should continue to loop. Returns false if
+// parent sent a message on channel indicating stop.
+func (devices *DeviceSet) waitSleepCleanupThread() bool {
+	// Create a timer of 30 seconds.
+	timer := time.NewTimer(time.Second * 30)
+
+	select {
+	case <-timer.C:
+		return true
+	case <-devices.deletionWorkerChan:
+		timer.Stop()
+	}
+	return false
+}
+
+// Don't call this function directly. Should be scheduled with a call to
+// scheduleDefDeletion()
+func (devices *DeviceSet) cleanupDeletedDevicesThread() {
+	devices.Lock()
+	for {
+		logrus.Debugf("devmapper: Calling cleanupDeletedDevices().\n")
+		err := devices.cleanupDeletedDevices()
+		if err != nil {
+			logrus.Errorf("devmapper: cleanupDeletedDevices() failed.")
+			break
+		}
+
+		if devices.nrDeletedDevices <= 0 {
+			break
+		}
+
+		devices.Unlock()
+		continueProcessing := devices.waitSleepCleanupThread()
+		devices.Lock()
+		if !continueProcessing {
+			break
+		}
+	}
+	devices.deletionWorkerRunning = false
+	logrus.Debugf("devmapper: Exiting cleanupDeletedDevicesThread.\n")
+	devices.Unlock()
+}
+
+// This should be called with devices.Lock() held.
+func (devices *DeviceSet) scheduleDefDeletion() {
+	if devices.deletionWorkerRunning {
+		return
+	}
+
+	logrus.Debugf("devmapper: Kickstarting deferred deletion worker.")
+	devices.deletionWorkerRunning = true
+	go devices.cleanupDeletedDevicesThread()
+}
+
+func (devices *DeviceSet) cancelDefDeletion() {
+	devices.Lock()
+	if devices.deletionWorkerRunning == false {
+		devices.Unlock()
+		return
+	}
+	devices.deletionWorkerChan <- true
+	devices.Unlock()
+
+	// Wait in a loop for thread to stop.
+	for {
+		devices.Lock()
+		if devices.deletionWorkerRunning == false {
+			devices.Unlock()
+			return
+		}
+		devices.Unlock()
+		// Sleep for 10 miliseconds and check again
+		time.Sleep(time.Millisecond * 10)
+	}
```

Comment:
If there is a bug, this can run indefinitely. Not sure what's the sane number where we can say stop. Thought it is not a big concern at this point as it is called during docker stop only and if Shutdown() does not finish in certain time (10 second), then all threads are killed anyway.